### PR TITLE
Disallowed --metadata-literal, --model-checker-show-unproved, --model-checker-div-mod-no-slacks outside of compiler mode

### PR DESCRIPTION
**Pull request by Midhun07**
Partially solves #11629.

Disallowed following options outside of compiler mode:  
1. --metadata-literal
2. --model-checker-show-unproved
3. --model-checker-div-mod-no-slacks 
4. . --via-ir
5. --metadata-hash=swarm

### CODE REVIEW

**Comment by Midhun07** [comment]
@cameel Added some more fixes please check.

File: Changelog.md
Comment (cameel) [inline Changelog.md]:
```suggestion
 * Commandline Interface: Disallow the following options outside of the compiler mode: ``--error-recovery``, ``--metadata-literal``, ``--model-checker-show-unproved``, ``--model-checker-div-mod-no-slacks``.
```

File: solc/CommandLineParser.cpp
Comment (cameel) [inline solc/CommandLineParser.cpp:915]: You could also take care of `--metadata-hash` option in the same PR. It's closely related and used in the same modes (i.e. compiler only).

File: solc/CommandLineParser.cpp
Comment (cameel) [inline solc/CommandLineParser.cpp]: Same here. All the `--model-checker-` options are closely related.

**Review by cameel — COMMENTED** [review]
Overall this looks good but please rebase on `develop` (I have just merged #11909). I thought you'll start a new branch on top of the one from #11909.

Also, changes are pretty simple at this point, so you could cover more options in one go. Especially the ones that are related.

**Comment by hrkrshnn** [comment]
@Midhun07 Thanks for the PR. Could you please follow up with the changes that cameel requested?

File: solc/CommandLineParser.cpp
Comment (Midhun07) [reply solc/CommandLineParser.cpp:915]: The reason why i left this option out is because it has a value assignment to it. So do we need to check that or simply checking if option is enabled or not is enough?

File: solc/CommandLineParser.cpp
Comment (Midhun07) [reply solc/CommandLineParser.cpp]: Same here.

**Comment by Midhun07** [comment]
@hrkrshnn @cameel  I have added changelog change. Other changes I need some help as in I just want to know how to check the values of the options. Like for example: `metadata_hash=swarm` How to check this option? Simple checking like other options is not working. Please help.

File: solc/CommandLineParser.cpp
Comment (cameel) [reply solc/CommandLineParser.cpp:915]: It's enough to check if the option is present (with `count()`). The check applies to the option as a whole, not specific values.

**Comment by cameel** [comment]
> Simple checking like other options is not working. Please help.

For most options `m_args.count(optionName) > 0` should work. For options that have default values and are always present because of this, you must check `!m_args[optionaName].defaulted()` instead.

**Comment by Midhun07** [comment]
@cameel My problem is that the message `The following options are not supported in the current input mode: --metadata-hash` is the error produced but the test compares that message against the message: `The following options are not supported in the current input mode: --metadata-hash=swarm`.  Is there any way to append `=swarm` apart from the usual string append? I mean something like `joinOptionNames`.

**Comment by cameel** [comment]
Well, you can do it in several ways:
- Split on `=` and remove everything after it when comparing.
- Have a more complex data structure. I.e. under `--metadata-hash` store both the allowed modes and a version with parameter value (`--metadata-hash=swarm`). Pass one into the command, use the other for comparison.
- Have two separate lists, one for options with values (and include the value there), one for options without.

**Comment by cameel** [comment]
I mean, you don't have to stick to the current stucture at all cost. You need to adjust it for this to work.

**Comment by hrkrshnn** [comment]
@Midhun07 Could you please rebase into develop instead of the merge? That is our workflow.

**Comment by Midhun07** [comment]
@hrkrshnn I have rebased my branch on top of `ethereum:develop`. Not sure though if it had worked properly. A little help would be much appreciated. Some 14 new commits have crept in. Not sure if this is the right way.

**Comment by cameel** [comment]
@Midhun07 I think you must have forgotten to pull in the latest commits into your local `develop` branch before rebasing.

**Comment by cameel** [comment]
Actually looks like it's something else. I'd have to look closer to see what actually happened but it was definitely now what should have happened :)  In any case it's the result of the branch not being properly rebased. If something like this happens and you're not sure what's wrong it's usually a good idea to visualize the branches. E.g. with `gitk --all` or `git log --all --graph`. The problem is usually easy to see.

Anyway, I rebased it properly and pushed just now.

File: solc/CommandLineParser.cpp
Comment (cameel) [inline solc/CommandLineParser.cpp:916]: I'd put `g_strMetadataHash` next to `g_strMetadataLiteral`.

File: test/solc/CommandLineParser.cpp
Comment (cameel) [inline test/solc/CommandLineParser.cpp]: Isn't it supposed to have a value on the command line?
```suggestion
			vector<string> commandLine = {"solc", optionName, "file", inputMode};
```

File: test/solc/CommandLineParser.cpp
Comment (cameel) [inline test/solc/CommandLineParser.cpp]:
```suggestion
			size_t separatorPosition = optionName.find("=");
```

File: test/solc/CommandLineParser.cpp
Comment (cameel) [inline test/solc/CommandLineParser.cpp]: `npos` in `substr()` represents the end of the string so you don't have to have a special case for it.
```suggestion
			string_view optionNameWithoutValue = optionName.substr(0, optionName.find("="));
```

Also, you can use `string_view` here to avoid copying the string.

Comment (cameel) [inline test/solc/CommandLineParser.cpp]: I'd add an assert against the name being empty here:
```c++
soltestAssert(!optionNameWithoutValue.empty());
```

Just to make sure that if someone by mistake uses `=` at the beginning of an option name on the list above, the code just stops instead of doing something weird (and potentially not even failing).

File: test/solc/CommandLineParser.cpp
Comment (Midhun07) [reply test/solc/CommandLineParser.cpp]: If we give the option as it is, it will throw an error as we are checking if the option is added or not in the `CommandLineParser`. That is why changed it to this.

Comment (cameel) [reply test/solc/CommandLineParser.cpp]: Not sure I understand. Are you saying that e.g. `solc --standard-json --metadata-hash=swarm` will report an error and `solc --standard-json --metadata-hash` won't? Isn't the error for `solc --standard-json --metadata-hash=swarm` exactly what should happen?

`solc  --standard-json --metadata-hash` does report an error but it's not the one we expect:
```
the required argument for option '--metadata-hash' is missing
```

Comment (Midhun07) [reply test/solc/CommandLineParser.cpp]: No that is not what I meant. In `CommandLineParser.cpp` we are checking if the option exists or not. There the string is `--metadata-hash` and not `--metadata-hash=swarm`. So in `test/CommandlineParser.cpp` we should be passing only `--metadata-hash` so that the test passes.

Comment (Midhun07) [reply test/solc/CommandLineParser.cpp]: I'm also asking if this is the right way to test this case. Since the option itself is invalid we need to check its value so I made it like this.

Comment (cameel) [reply test/solc/CommandLineParser.cpp]: If I understand correctly, you're worried that `CommandLineParser` will see `--metadata-hash=swarm` as a single option named `metadata-hash=swarm` rather than `metadata-hash` option with a value of `swarm`? That should not be the case. boost::program_options handles both `--metadata-hash swarm` and `--metadata-hash=swarm` in the same way. So if you call `solc --metadata-hash=swarm`, `m_args.count("metadata-hash")` should still give you 1.

File: solc/CommandLineParser.cpp
Comment (cameel) [reply solc/CommandLineParser.cpp]: So what about this? Can you add all model checker options? I think we should do them all in one go.

**Comment by cameel** [comment]
This needed a rebase and had some small issues (tabs in the middle of a line in a few places, not all options listed in the changelog. Also we have just released 0.8.10 so the changelog entry should go to 0.8.11 now.

I fixed these and pushed my changes to your branch.

Not all comments above are resolved so I cannot approve yet. Please take a look at them and let me know if any of them is problematic.

**Review by dzhalaevd — DISMISSED** [review]

File: test/solc/CommandLineParser.cpp
Comment (Midhun07) [reply test/solc/CommandLineParser.cpp]: This is removed.

**Comment by Midhun07** [comment]
Hi @cameel, I have been trying to implement the suggestions. The `string_view` isn't working fine. I'm not able to reconstruct the original string back. It is printing garbage. Here is what I'm trying to do.

`string_view optionNameWithoutValue = optionName.substr(0, optionName.find("="));`
this is to get the address and the length of the substring. And then to reconstruct it:
`string(optionNameWithoutValue.begin(), optionNameWithoutValue.length())`.

I have tried this on a template and there it was working fine. For some reason it is not working here. Is there any issue with this reconstruction? Also I have been trying to debug using `soltest.sh --debug`. But the working of the test cases is little different from a normal execution i think. I'm not able to print the value of the variables.

**Comment by cameel** [comment]
@Midhun07 Try `string(optionNameWithoutValue)`.

But I'm really surprised that the operator for adding a `string_view` to `string` is not available in C++. Apparently it's by design: [Why is there no support for concatenating std::string and std::string_view?](https://stackoverflow.com/questions/44636549/why-is-there-no-support-for-concatenating-stdstring-and-stdstring-view/47735624#47735624).

This makes this change pointless so, well, let's go back to `string`.

> Also I have been trying to debug using `soltest.sh --debug`. But the working of the test cases is little different from a normal execution i think. I'm not able to print the value of the variables.

It's debuggable just as `solc`. I just checked and I can actually inspect variables. The only ones I can't are the "fake" ones that come from destructuring (`optionName` and `inputModes`). Maybe you were trying to inspect these?

**Comment by leonardoalt** [comment]
This PR has been stale for a while. Maybe take over?

**Comment by cameel** [comment]
@nishant-sachdeva 
> Slight doubt regarding git. 
> I took over the PR https://github.com/Midhun07/solidity/tree/fix_invalid_options_mode_combinations.
> Have rebased i and  pushed changes to it
> But that is not being reflected in the PR created from this branch at https://github.com/ethereum/solidity/pull/12166
 >
> So, the question is, could I be missing a step here? Any inputs would be great!

Are you sure you're pushing to the right remote? This is an extrnal PR so you need to add the contributor's fork as a remote first:
```bash
git remote add Midhun07 git@github.com:Midhun07/solidity.git
```
Then you can do `git push Midhun07 fix_invalid_options_mode_combinations --force-with-lease`

**Comment by cameel** [comment]
By the way I see that you did manage to push here. But something went very wrong because there's a ton of extra commits. Is the branch properly rebased on `develop`? Also make sure it was rebased, not merged. There should be no merge commits.

File: solc/CommandLineParser.cpp
Comment (cameel) [inline solc/CommandLineParser.cpp]: Looks like some unresolved conflicts were committed.

@nishant-sachdeva Will you be continuing this? At least this should be cleaned up but I see you unassigned yourself.

File: test/solc/CommandLineParser.cpp
Comment (nishant-sachdeva) [reply test/solc/CommandLineParser.cpp]: Marking this as resolved, since in a later comment, it's been advised to abandon the use of `string_view`

File: test/solc/CommandLineParser.cpp
Comment (cameel) [inline test/solc/CommandLineParser.cpp]: @nishant-sachdeva 
> Can't quite figure out how to fix the Boost tests that are failing.
> Would love some pointers on that

`BOOST_TEST` here is failing because stderr is empty. And it's right. `parseCommandLine()` is supposed to throw an exception, not print to stderr and `BOOST_CHECK_EXCEPTION()` already correctly checks for that. The `BOOST_TEST` line should not be here at all.

Also, the error message you're getting is a bit confusing so let's look at it closer:
```
message: check serr.str() == "The following options are not supported in the current input mode: " + optionNameWithoutValue + "\n" has failed [ != The following options are not supported in the current input mode: --error-recovery
]
```
What it's saying is that `""` is not equal to `"The following options are not supported in the current input mode: --error-recovery"`. It's just that there are no quotes around the strings so it looks as if the empty string were not there and it were comparing with `"The following options are not supported in the current input mode: " + optionNameWithoutValue + "\n"` instead.

File: solc/CommandLineParser.cpp
Comment (cameel) [reply solc/CommandLineParser.cpp]: @nishant-sachdeva 
> Since some of the model-checker options have different values for the flags, and we may not need to disallow all of the values, is there some compiled list / documentation which mentions what options are allowed for which compiler mode ?
> Or do have to manually go through the code and check that ?

If you want the values allowed on the CLI see `solc --help` or [this part of `CommandLineParser::optionsDescription()`](https://github.com/ethereum/solidity/blob/v0.8.14/solc/CommandLineParser.cpp#L785-L834).

Same info for Standard JSON is under [Compiler Input and Output JSON Description > Input description](https://docs.soliditylang.org/en/latest/using-the-compiler.html#input-description). Look at the `modelChecker` field.

And [SMTChecker and Formal Verification](https://docs.soliditylang.org/en/latest/smtchecker.html) has an in-depth explanation of what these options do.

File: test/solc/CommandLineParser.cpp
Comment (nishant-sachdeva) [reply test/solc/CommandLineParser.cpp]: I see. 
I'll remove the BOOST_TEST line and push changes.

File: solc/CommandLineParser.cpp
Comment (cameel) [reply solc/CommandLineParser.cpp]: @nishant-sachdeva This one is still unresolved. Can you add all the model checker options?

File: Changelog.md
Comment (cameel) [inline Changelog.md]: `--experimental-via-ir` and `--error-recovery` were already covered in previous PRs.
```suggestion
 * Commandline Interface: Disallow the following options outside of the compiler mode: ``--via-ir``, ``--metadata-literal``, ``--metadata-hash``, ``--model-checker-show-unproved``, ``--model-checker-div-mod-no-slacks``.
```

File: test/solc/CommandLineParser.cpp
Comment (cameel) [reply test/solc/CommandLineParser.cpp]: @nishant-sachdeva I think this still needs to be changed.

File: solc/CommandLineParser.cpp
Comment (nishant-sachdeva) [reply solc/CommandLineParser.cpp]: Pushed code to resolve this comment.

File: test/solc/CommandLineParser.cpp
Comment (nishant-sachdeva) [reply test/solc/CommandLineParser.cpp]: Added this change in a later push to the code.

File: Changelog.md
Comment (cameel) [inline Changelog.md]: Needs to be moved to 0.8.16 now.

File: solc/CommandLineParser.cpp
Comment (cameel) [inline solc/CommandLineParser.cpp]:
```suggestion
			m_args.count(optionName) > 0 &&
			inputModes.count(m_options.input.mode) == 0 &&
			!m_args[optionName].defaulted()
```

**Review by cameel — DISMISSED** [review]
Looks good now apart from some minor style issues. I'm approving because I won't be available later but we do still need to move the changelog entry before we merge.

**Review by ekpyron — APPROVED** [review]
Reapproving on a casual look only, based on @cameel's earlier approval.

**Pull request merged**
